### ISSUE TITLE
Dispose doesnt work when page is refreshed (server-side blazor)

### ISSUE THREAD

**Issue by FeeSaver** [aspnet/Blazor#1648]
Hello I was trying to make the fetchdata example update values without refreshing page or clicking buttons. And I found out that I can use timer which seemed like working ok. But after testing with different data it was acting strange and I found out that page doesnt dispose the timer when I go to different page and its not displayed anymore. 
That was solved by adding dispose but it didnt solve different issue.. When I press F5 while on /fetchdata page it reloads and doesnt dispose the timer. And then it runs twice. I tried it by adding int into ForecastService and it increment everytime GetForecastAsync is called. After few 10 refreshes it increment +10 every second instead of +1. 
I hope it makes sense. 

```
@functions {
WeatherForecast[] forecasts;
Timer timer;
private async void Update()
{
    forecasts = await ForecastService.GetForecastAsync(DateTime.Now);
    StateHasChanged();
}

protected override async Task OnInitAsync()
{
    timer = new Timer(1000);
    timer.Elapsed += (sender, args) => Update();
    timer.Start();
    Update();
}

public void Dispose()
{
    timer.Dispose();
}
}
```

**Comment by RyoukoKonpaku** [comment]
It seems a bug indeed from what I can tell it seems it couldn't update the server from SignalR that a page refresh happened to properly dispose the component. You can tell this by turning-on debug and setting a break point on the dispose method. It hits when you go to a different route but not when a refresh happens.

It would probably work if you could manually dispose the component before a refresh triggers via `windows.beforeunload` event using `JSInterop` though a more integrated approach from the framework itself to would be ideal. Probably by informing the server a refresh happened.

**Comment by SteveSandersonMS** [comment]
Yes, this is one of several areas of cleanup and resource-usage resilience we know we need to implement in the server-side model. Thanks for raising this - I'll leave it open until we can get to it (which hopefully will be soon).

**Comment by chrdlx** [comment]
I could replicate this too... Also server-side.
After hours of trying to find a bug in my event handlers I found that I was subscribing to events from blazor components but never unsuscribing due to the Dispose not being called ever...  glad to found here  that it was just a bug  =)
